### PR TITLE
Handle late clock trigger task event

### LINKED ISSUE #2207: Alerting that a clock-triggered task is delayed?
It might be useful to alert users that a clock-triggered task's time has arrived, but the task cannot trigger due to other dependencies not being satisfied. 

The motivation behind this is, operational people like to know if things are running "late".  NIWA operations currently uses Nagios IT infrastructure monitoring for this sort of thing - certain tasks are expected to have reported in to Nagios by certain times.  I figure we could do it in Cylc by alerting when a clock-trigger is up but the task can't run for other reasons, because the clock-trigger represents the time that the suite writer thinks the task *should* run.

### PR DESCRIPTION
**Pull request by matthewrmshin**
Log and emit a *late* event when a task reaches its clock trigger time ~but has unmet prerequisites~.

~(In this implementation, it will only emit the event as the suite runs, so tasks with clock-trigger before the time of current suite start up (or restart) will not get reported. Happy to improve on this behaviour.)~

~In the latest implementation, the suite will log and emit a late event *once* for each clock trigger task that is late for more than a minute. This should work correctly on start up and restart.~

In the latest implementation, the suite will log and emit a late event *once* for each task that has `late offset` configured and that the current time is beyond its cycle point time + the late offset. This should work correctly on start up and restart.

Close #2207. Supersede #2594.

### CODE REVIEW

**Comment by hjoliver** [comment]
> (In this implementation, it will only emit the event as the suite runs, so tasks with clock-trigger before the time of current suite start up (or restart) will not get reported. Happy to improve on this behaviour.)

I can see how your use of `prev_late_time_check = now` is a nice clean way to do this, but IMO tasks should emit a late event even if "clock trigger time" is not greater than "now" - after all such tasks are even "later" than those that can't run at their clock-trigger times once the suite has caught up. (Does that make sense?!)

**Comment by matthewrmshin** [comment]
Should be easy enough to implement if I understand this correctly.
* Report late tasks on suite start up?
* Remember previous late check time on restart?

**Comment by hjoliver** [comment]
No, what I mean is easier than that: every task proxy should emit a late event if the wall-clock time is greater than its clock-trigger time when it first checks.

Future instances of the same task may also be late already, but their task proxies don't exist yet so they can't check yet.  But I don't think that matters.  Late events will come from the most recent instance of a task that could potentially run now given the progress of the suite to date.

I think this implies each task has to remember if it has emitted a late check yet, and maybe that should persist across restart.

**Comment by matthewrmshin** [comment]
OK. I think I get what you mean.

**Comment by hjoliver** [comment]
Maybe we still haven't understood each other on what exactly these late events should be, but this doesn't work as I expected.  You global prev_late_time check results in only a single late event being emitted for all late tasks.  Taking the simplest possible clock-triggered suite as an example:
```
[scheduling]
  initial cycle point = 2015
  [[special tasks]]
      clock-trigger = foo(PT0H)
  [[dependencies]]
      [[[P1Y]]]
          graph = "foo[-P1Y] => foo"
[runtime]
   [[foo]]
      script = sleep 10
      [[[events]]]
          late handler = "echo !!!LATE"
```

What I'd expect is, tasks foo.2015 through foo.2018 should each emit a late event shortly after they each come into existence, then foo.2019 onward will submit on time with no further late events emitted.
This seems the most sensible way of late-alerting to me (even though foo.2016, say, has technically been late ever since Jan 2016, its lateness could not be alerted before now because the suite was not even running until now ... or because the suite was running so far behind that it had not become aware of .2016 tasks until now).
 
But on this branch, only foo.2015 emits a late event.

**Comment by matthewrmshin** [comment]
That's not the intention, but you are right about the `prev_late_time` flag. It is because `foo.2016..8` are not yet spawned when we first run through the main loop, so the global `prev_late_time` has moved from epoch to a much more current time. My intention is to avoid adding another table/column to the runtime database (which can cause compatibility issues), but perhaps this is unavoidable.

**Comment by matthewrmshin** [comment]
Latest implementation should do it!

**Comment by hjoliver** [comment]
That's got it.  Thanks.

**Comment by hjoliver** [comment]
@matthewrmshin - I've made an attempt at documenting this feature, which I think is not going to be easy to explain to users... I'll post a PR to your branch later tonight.

**Comment by sadielbartholomew** [comment]
I have one general comment (I'm starting to look at the code & will submit a review this afternoon). This is currently being implemented in a 'one-size-fits-all' way, but given different users (ops vs. research vs. non-meteorological etc) will surely naturally to treat 'late' events with different levels of urgency would it be worth the effort to include a small level of customisability here?

I feel it would be useful, & easy enough, to make the 'threshold' time past the clock trigger at which the event is handled (that is currently set to 1 min, a sensible default value I think) changeable via a late handler 'option' or similar to some preferred time (to include zero), & also to override the logging level to 'INFO' as for some always logging as a 'WARNING' could get irritating &/or clog up their log viewer so that contextually more important warnings become hard to spot or missed.

If you think this generally a good idea, we could also consider implementing the means to disable the characteristic discussed here & in the original issue whereby tasks with clock triggers subsequent to suite start-up (or restart) time emit late events. It does seem to me that last year when it was originally discussed everyone could at least see the merits of 'historical' tasks *not* emitting, though we now want them to; if there could be a significant divide in preference for this why not effectively provide a choice?

**Comment by hjoliver** [comment]
@sadielbartholomew - I'll just weigh in on your suggestions above.

 * Personally I'm happy with "warning" severity for late events - it's not a normal event, but it's not critical either.  Configurability that's not really needed just added complexity.
 * Making the threshold configurable is probably a good idea because clock-triggers are typically set at the earliest time that a task should run, and 1 minute after that is probably not a big deal.   However, I'm about to suggest a more fundamental change below that would make this unnecessary.
 * Disabling late events for tasks whose cycle points pre-date the suite restart time might be sensible, and maybe we should do it (I'd need to think more on this...) but in any case it's a minor issue because late alerting is really only needed in operational contexts, and operational suites (almost?) never get more than a whole cycle behind.

**Comment by hjoliver** [comment]
**problem!**

@cylc/core - if you read the additions to `cug.tex` this PR (which came to @matthewrmshin from me) you'll see that while trying to explain how to use this feature, I more or less convinced myself that it is useless and ended recommending use of an external monitoring system to determine lateness (which is what we already do at NIWA).  That's because:
 * Cylc can only absolutely identify lateness in clock-triggered tasks,
 * but clock-triggered tasks are typically not the ones that need late alerting
 * and even if they were the ones, they would typically not generate late events during typical operational delays of less than one cycle, because they typically have no upstream dependencies and can therefore always run on time (so long as the task proxy exists already).
 * and finally (as explained in cug.tex) it is a bad idea to put artificial clock-triggers on the "important" tasks, or to put artificial dependencies on the natural clock-triggered tasks.
... therefore, we cannot get useful late alerting out of this new feature as implemented.  (apologies @matthewrmshin for failing to realise this before you did the implementation!  - however, most of your code will work for what follows...)

File: lib/cylc/task_state.py [hunk 1]
```diff
@@ -96,6 +96,16 @@
     TASK_STATUS_SUBMIT_FAILED,
 ])
 
+# Task statuses that are never active.
+# For tasks that have never been submitted, but excluding expired and held.
+TASK_STATUSES_NEVER_ACTIVE = set([
+    TASK_STATUS_RUNAHEAD,
+    TASK_STATUS_WAITING,
+    TASK_STATUS_HELD,
```
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/task_state.py]: The `held` state is almost always imposed manually, so their lateness should not be reported until they are released.

File: lib/cylc/suite_db_mgr.py [hunk 2]
```diff
@@ -236,32 +238,28 @@ def put_runtime_inheritance(self, config):
...
         """
         self.db_inserts_map[self.TABLE_SUITE_PARAMS].extend([
-            {"key": "run_mode", "value": run_mode},
-            {"key": "cylc_version", "value": cylc_version},
-            {"key": "UTC_mode", "value": UTC_mode},
-            {"key": "initial_point", "value": str(initial_point)},
-            {"key": "final_point", "value": str(final_point)},
+            {"key": "run_mode", "value": schd.run_mode},
+            {"key": "cylc_version", "value": CYLC_VERSION},
+            {"key": "UTC_mode", "value": cylc.flags.utc},
```
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/suite_db_mgr.py]: SQLite does not have a native boolean type. In such case, it is normally best to represent boolean data using the integers `0|1`, which can be used in boolean context without any conversion. (The expression `bool('False')` evaluates to `True` in Python, whereas `bool(0)` evaluates to `False`.)

**Comment by matthewrmshin** [comment]
Should be easy enough to implement.

File: lib/cylc/task_state.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hjoliver) [reply lib/cylc/task_state.py]: That's a good point that is not entirely obvious - can you add a comment for future reference?

**Comment by sadielbartholomew** [comment]
Thanks for your response to my questions, @hjoliver, & sorry for not picking up on the fact that you were had reservations about the lack of use of the late event in the ``cug.tex`` commit which was merged into this PR shortly before I commented, I skim read it & took it as advice that these late events should be used sparingly.

Absolutely agree that 'configurability that's not really needed just adds unnecessary complexity' & these new points regarding usefulness of the current idea are very significant so glad you brought them up. I have little knowledge of practical operational system management but your solution seems very sensible.

**Comment by hjoliver** [comment]
> sorry for not picking up on the fact that you were had reservations...

That's alright - I didn't really pick up on the implications of what I was saying myself, until the next morning, after I'd slept on it!

File: lib/cylc/task_state.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply lib/cylc/task_state.py]: But surely then the ``TASK_STATUS_HELD`` entry should be omitted from this set, as the ``late_tasks_check()`` function tests whether a task has a state belonging to the set (``scheduler.py`` current line 1214) as part of the conditional for establishing 'lateness'? I assume from the UG that releasing a 'held' task will always revert the state to some non-active one i.e. one already in this set, so to not report until release would require its exclusion. Not sure what I have missed here!

File: lib/cylc/suite_db_mgr.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply lib/cylc/suite_db_mgr.py]: Thanks for explaining :)

**Comment by matthewrmshin** [comment]
Lateness is no longer related to clock trigger. A `late offset` setting is used to determine if a task is late compared to its cycle point or not. (Or should it be called `waiting timeout`, like `submission timeout and `execution timeout`? ) :watch: :hourglass:

**Comment by matthewrmshin** [comment]
I'll add something to the CUG when we can finally agree with the approach.

File: lib/cylc/scheduler.py [hunk 3]
```diff
@@ -1227,6 +1207,22 @@ def database_health_check(self):
             # Something has to be very wrong here, so stop the suite
             raise SchedulerError(str(exc))
 
+    def late_tasks_check(self):
+        """Report tasks that are late for their clock triggers."""
```
Comment (hjoliver) [inline lib/cylc/scheduler.py]: doc-string is now wrong

**Comment by hjoliver** [comment]
> I'll add something to the CUG when we can finally agree with the approach.

I agree with this approach (not surprisingly perhaps!), so we're good to go if you and your team agree.

We should keep much of the cug documentation already added, to explain why lateness can't easily  be determined automatically. ... I'll have a go at quickly reformulating the docs now and put up a PR to your branch.

Also I'm happy with the term "late offset" because unlike the timeouts, it is relative to cycle point not a timeout after some interval with a well-defined start point.

**Comment by sadielbartholomew** [comment]
I also agree with the approach & with 'late offset' as the descriptor (seems intuitive). Let me know when this is ready for re-review.

File: lib/cylc/task_proxy.py [hunk 4]
```diff
@@ -24,32 +24,155 @@
...
-                 "point", "cleanup_cutoff", "identity", "has_spawned",
-                 "point_as_seconds", "stop_point", "manual_trigger",
-                 "is_manual_submit", "summary", "local_job_file_path",
-                 "try_timers", "task_host", "task_owner",
-                 "job_vacated", "poll_timers", "timeout_timers",
-                 "delayed_start", "expire_time", "state"]
+    __slots__ = [
+        'cleanup_cutoff',
+        'clock_trigger_time',
+        'expire_time',
```
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/task_proxy.py]: I think it was missed in the description. The attribute is still used.

File: lib/cylc/scheduler.py [hunk 5]
```diff
@@ -1044,17 +1029,17 @@ def load_suiterc(self, is_reload=False):
...
         """
         key, value = row
-        if key == 'initial_point':
-            self._cli_initial_point_string = value
+        if key == "initial_point":
+            self.cli_initial_point_string = value
             self.task_events_mgr.pflag = True
-        elif key == 'warm_point':
-            self._cli_start_point_string = value
+        elif key in ["start_point", "warm_point"]:
```
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/scheduler.py]: For back compat. Comment added.

**Comment by matthewrmshin** [comment]
All comments addressed.

File: lib/cylc/task_proxy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (matthewrmshin) [reply lib/cylc/task_proxy.py]: Attribute now added to doc string.

**Pull request merged**